src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -3786,92 +3786,59 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
-                    // is cheaper than doing bitmask + popcount on the full last vector. To avoid complicated type
-                    // based checks, other remainder-count based logic to determine the correct cut-off, for simplicity
-                    // a half-vector size is chosen (based on benchmarks).
-                    uint remaining = (uint)Unsafe.ByteOffset(ref current, ref end) / (uint)Unsafe.SizeOf<T>();
-                    if (remaining > Vector512<T>.Count / 2)
-                    {
-                        ulong mask = Vector512.Equals(Vector512.LoadUnsafe(ref oneVectorAwayFromEnd), targetVector).ExtractMostSignificantBits();
 
-                        // The mask contains some elements that may be double-checked, so shift them away in order to get the correct pop-count.
-                        uint overlaps = (uint)Vector512<T>.Count - remaining;
-                        mask >>= (int)overlaps;
-                        count += BitOperations.PopCount(mask);
-
-                        return count;
-                    }
+                    // Count the last vector and mask off the elements that were already counted (number of elements between oneVectorAwayFromEnd and current).
+                    ulong mask = Vector512.Equals(Vector512.LoadUnsafe(ref oneVectorAwayFromEnd), targetVector).ExtractMostSignificantBits();
+                    mask >>= (int)((nuint)Unsafe.ByteOffset(ref oneVectorAwayFromEnd, ref current) / (uint)Unsafe.SizeOf<T>());
+                    count += BitOperations.PopCount(mask);
                 }
                 else if (Vector256.IsHardwareAccelerated && length >= Vector256<T>.Count)
                 {
                     Vector256<T> targetVector = Vector256.Create(value);
                     ref T oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector256<T>.Count);
-                    do
+                    while (Unsafe.IsAddressLessThan(ref current, ref oneVectorAwayFromEnd))
                     {
                         count += BitOperations.PopCount(Vector256.Equals(Vector256.LoadUnsafe(ref current), targetVector).ExtractMostSignificantBits());
                         current = ref Unsafe.Add(ref current, Vector256<T>.Count);
                     }
-                    while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd));
 
-                    // If there are just a few elements remaining, then processing these elements by the scalar loop
-                    // is cheaper than doing bitmask + popcount on the full last vector. To avoid complicated type
-                    // based checks, other remainder-count based logic to determine the correct cut-off, for simplicity
-                    // a half-vector size is chosen (based on benchmarks).
-                    uint remaining = (uint)Unsafe.ByteOffset(ref current, ref end) / (uint)Unsafe.SizeOf<T>();
-                    if (remaining > Vector256<T>.Count / 2)
```

Comment:
Are there other APIs where we've done a similar guard, or just with Count?

Reply:
I don't remember seeing a similar pattern where we'd revert to the scalar loop depending on how much data remains. I'm not immediately seeing anything too similar on Grep.app.

The closest thing that comes to mind is this logic in ProbabilisticMap
https://github.com/dotnet/runtime/blob/24b59d85541c72fd55b8d5d0bc9adb84619d6c4f/src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticMap.cs#L528-L540
where we switch between a final iteration of Vector128 (preferred) vs. a final iteration of Vector256.
But I did that there because it's possible the already-handled (overlapping) elements were false-positive matches we had to rule out one-by-one in the verification step, so the potential overhead of having to recheck them is a lot higher than here in `Count` where it's practically free.

Reply:
> Are there other APIs where we've done a similar guard, or just with Count?

I also think this one was the only place where we did this.